rust/connlib/tunnel/src/lib.rs

```diff
@@ -132,52 +147,72 @@ where
     }
 }
 
-impl<CB> Tunnel<CB, GatewayState, Server, ClientId>
+impl<CB> GatewayTunnel<CB>
 where
     CB: Callbacks + 'static,
 {
-    pub fn poll_next_event(&mut self, cx: &mut Context<'_>) -> Poll<Result<Event<ClientId>>> {
-        match self.role_state.poll(cx) {
-            Poll::Ready(()) => {
-                cx.waker().wake_by_ref();
+    pub fn new(private_key: StaticSecret, callbacks: CB) -> Result<Self> {
+        let now = Instant::now();
+
+        Ok(Self {
+            callbacks: CallbackErrorFacade(callbacks),
+            state: GatewayState::new(private_key, now),
+            device: None,
+            no_device_waker: AtomicWaker::new(),
+            timeout: None,
+            sockets: Sockets::new()?,
+            write_buf: Box::new([0; MAX_UDP_SIZE]),
+            read_buf: Box::new([0; MAX_UDP_SIZE]),
+        })
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<gateway::Event>> {
+        if let Some(timeout) = self.state.poll_timeout() {
+            let timeout = tokio::time::Instant::from_std(timeout);
+
+            match self.timeout.as_mut() {
+                Some(existing_timeout) if existing_timeout.deadline() != timeout => {
+                    existing_timeout.as_mut().reset(timeout)
+                }
+                Some(_) => {}
+                None => self.timeout = Some(Box::pin(tokio::time::sleep_until(timeout))),
             }
-            Poll::Pending => {}
+        }
+
+        if let Some(timeout) = self.timeout.as_mut() {
+            ready!(timeout.poll_unpin(cx));
+            self.state.handle_timeout(timeout.deadline().into());
+
+            cx.waker().wake_by_ref()
+        }
```

Comment:
@conectado Most of this could be extracted into a `MaybeTimeout` type that can be reused between a `GatewayTunnel` and a `ClientTunnel`.